specs/_features/eip7594/fork-choice.md

```diff
@@ -0,0 +1,191 @@
+a custody check, while for blocks older than two epochs through a full peer sampling check.
+
+```python
+def get_head(store: Store) -> Root:
+    # Get filtered block tree that only includes viable branches
+    blocks = get_filtered_block_tree(store)
+    # Execute the LMD-GHOST fork choice
+    head = store.justified_checkpoint.root
+    while True:
+        # Get available children for the current slot
+        children = [
+            root for (root, block) in blocks.items()
+            if (
+                block.parent_root == head
+                and is_data_available(
+                    root, 
+                    require_peer_sampling=is_peer_sampling_required(store, block.slot)
+                )
+            )
+        ]
+        if len(children) == 0:
+            return head
+        # Sort by latest attesting balance with ties broken lexicographically
+        # Ties broken by favoring block with lexicographically higher root
+        head = max(children, key=lambda root: (get_weight(store, root), root))
+```
+
+#### New `is_peer_sampling_required`
+
+```python
+def is_peer_sampling_required(store: Store, slot: Slot):
+    return compute_epoch_at_slot(slot) + 2 <= get_current_store_epoch(store)
+```
+
+## Updated fork-choice handlers
+
+### Modified `on_block`
+
+*Note*: The blob data availability check is removed and replaced with an availability
+check on the justified checkpoint in the "pulled up state" of the block, which is
+the state after applying `process_justification_and_finalization`.
+
+```python
+def on_block(store: Store, signed_block: SignedBeaconBlock) -> None:
+    """
+    Run ``on_block`` upon receiving a new block.
+    """
+    block = signed_block.message
+    # Parent block must be known
+    assert block.parent_root in store.block_states
+    # Make a copy of the state to avoid mutability issues
+    state = copy(store.block_states[block.parent_root])
+    # Blocks cannot be in the future. If they are, their consideration must be delayed until they are in the past.
+    assert get_current_slot(store) >= block.slot
+
+    # Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)
+    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
+    assert block.slot > finalized_slot
+    # Check block is a descendant of the finalized block at the checkpoint finalized slot
+    finalized_checkpoint_block = get_checkpoint_block(
+        store,
+        block.parent_root,
+        store.finalized_checkpoint.epoch,
+    )
+    assert store.finalized_checkpoint.root == finalized_checkpoint_block
+
+    # Check the block is valid and compute the post-state
+    block_root = hash_tree_root(block)
+    state_transition(state, signed_block, True)
+
+    # [New in EIP7594] Do not import the block if its unrealized justified checkpoint is not available
+    pulled_up_state = state.copy()
```

Comment:
I cannot see any place in the rest of this function where `state` is modified.
If so, then we don't need to create a copy of it.